zerver/tests/test_audit_log.py

```diff
@@ -264,3 +283,179 @@ def test_get_last_message_id(self) -> None:
+            })).count(), 1)
+
+    def test_change_icon_source(self) -> None:
+        now = timezone_now()
+        realm = get_realm('zulip')
+        user = self.example_user('hamlet')
+        icon_source = 'G'
+        do_change_icon_source(realm, icon_source, acting_user=user)
+        self.assertEqual(RealmAuditLog.objects.filter(realm=realm, event_type=RealmAuditLog.REALM_ICON_SOURCE_CHANGED,
+                                                      acting_user=user, event_time__gte=now).count(), 1)
+        self.assertEqual(icon_source, realm.icon_source)
+
+    def test_change_subscription_property(self) -> None:
+        user = self.example_user('hamlet')
+        # Fetch the Denmark stream for testing
+        stream = get_stream("Denmark", user.realm)
+        sub = Subscription.objects.get(user_profile=user, recipient__type=Recipient.STREAM,
+                                       recipient__type_id=stream.id)
+        properties = {"color": True,
+                      "in_home_view": False,
+                      "is_muted": True,
+                      "desktop_notifications": False,
+                      "audible_notifications": False,
+                      "push_notifications": True,
+                      "email_notifications": True,
+                      "pin_to_top": True,
+                      "wildcard_mentions_notify": False}
+
+        for property, value in properties.items():
+            now = timezone_now()
+            if property == "in_home_view":
+                property = "is_muted"
+
+            old_value = getattr(sub, property)
+            do_change_subscription_property(user, sub, stream, property, value, acting_user=user)
+            expected_extra_data = {RealmAuditLog.OLD_VALUE: {'property': property, 'value': old_value},
+                                   RealmAuditLog.NEW_VALUE: {'property': property, 'value': value}}
+            self.assertEqual(RealmAuditLog.objects.filter(
+                realm=user.realm, event_type=RealmAuditLog.SUBSCRIPTION_PROPERTY_CHANGED,
+                event_time__gte=now, acting_user=user, modified_user=user,
+                extra_data=ujson.dumps(expected_extra_data)).count(), 1)
+            self.assertEqual(getattr(sub, property), value)
+
+    def test_change_default_streams(self) -> None:
+        now = timezone_now()
+        user = self.example_user('hamlet')
+        stream = get_stream("Denmark", user.realm)
+
+        old_value = getattr(user, 'default_sending_stream')
+        do_change_default_sending_stream(user, stream, acting_user=user)
+        self.assertEqual(RealmAuditLog.objects.filter(
+            realm=user.realm, event_type=RealmAuditLog.USER_DEFAULT_SENDING_STREAM_CHANGED,
+            event_time__gte=now, acting_user=user,
+            extra_data=ujson.dumps({
+                RealmAuditLog.OLD_VALUE: old_value,
+                RealmAuditLog.NEW_VALUE: stream
```

Comment:
Uh, JSON encoding this `Stream` object does not do what you think:

```json
{
  "API_FIELDS":[
    "name",
    "id",
    "description",
    "rendered_description",
    "invite_only",
    "is_web_public",
    "stream_post_policy",
    "history_public_to_subscribers",
    "first_message_id",
    "message_retention_days",
    "date_created"
  ],
  "MAX_DESCRIPTION_LENGTH":1024,
  "MAX_NAME_LENGTH":60,
  "MESSAGE_RETENTION_SPECIAL_VALUES_MAP":{
    "forever":-1,
    "realm_default":null
  },
  "STREAM_POST_POLICY_ADMINS":2,
  "STREAM_POST_POLICY_EVERYONE":1,
  "STREAM_POST_POLICY_RESTRICT_NEW_MEMBERS":3,
  "STREAM_POST_POLICY_TYPES":[
    1,
    2,
    3
  ],
  "date_created":1596760684,
  "deactivated":false,
  "description":"A Scandinavian country",
  "email_token":"01469f12695a36d617c727b3ffe7c386",
  "first_message_id":5,
  "history_public_to_subscribers":true,
  "id":1,
  "invite_only":false,
  "is_in_zephyr_realm":false,
  "is_web_public":false,
  "message_retention_days":null,
  "name":"Denmark",
  "pk":1,
  "realm":{
    "AUTHENTICATION_FLAGS":[
      "Google",
      "Email",
      "GitHub",
      "LDAP",
      "Dev",
      "RemoteUser",
      "AzureAD",
      "SAML",
      "GitLab",
      "Apple"
    ],
    "BOT_CREATION_ADMINS_ONLY":3,
    "BOT_CREATION_EVERYONE":1,
    "BOT_CREATION_LIMIT_GENERIC_BOTS":2,
    "BOT_CREATION_POLICY_TYPES":[
      1,
      2,
      3
    ],
    "COMMON_POLICY_TYPES":[
      1,
      2,
      3
    ],
    "COMMUNITY":2,
    "CORPORATE":1,
    "DEFAULT_COMMUNITY_TOPIC_EDITING_LIMIT_SECONDS":86400,
    "DEFAULT_MESSAGE_CONTENT_DELETE_LIMIT_SECONDS":600,
    "DEFAULT_MESSAGE_CONTENT_EDIT_LIMIT_SECONDS":600,
    "DEFAULT_NOTIFICATION_STREAM_NAME":"general",
    "DIGEST_WEEKDAY_VALUES":[
      0,
      1,
      2,
      3,
      4,
      5,
      6
    ],
    "EMAIL_ADDRESS_VISIBILITY_ADMINS":3,
    "EMAIL_ADDRESS_VISIBILITY_EVERYONE":1,
    "EMAIL_ADDRESS_VISIBILITY_MEMBERS":2,
    "EMAIL_ADDRESS_VISIBILITY_NOBODY":4,
    "EMAIL_ADDRESS_VISIBILITY_TYPES":[
      1,
      3,
      4
    ],
    "ICON_FROM_GRAVATAR":"G",
    "ICON_SOURCES":[
      [
        "G",
        "Hosted by Gravatar"
      ],
      [
        "U",
        "Uploaded by administrator"
      ]
    ],
    "ICON_UPLOADED":"U",
    "INITIAL_PRIVATE_STREAM_NAME":"core team",
    "INVITES_STANDARD_REALM_DAILY_MAX":3000,
    "LIMITED":2,
    "LOGO_DEFAULT":"D",
    "LOGO_SOURCES":[
      [
        "D",
        "Default to Zulip"
      ],
      [
        "U",
        "Uploaded by administrator"
      ]
    ],
    "LOGO_UPLOADED":"U",
    "MAX_REALM_NAME_LENGTH":40,
    "MAX_REALM_SUBDOMAIN_LENGTH":40,
    "MESSAGE_RETENTION_SPECIAL_VALUES_MAP":{
      "forever":-1
    },
    "MESSAGE_VISIBILITY_LIMITED":10000,
    "POLICY_ADMINS_ONLY":2,
    "POLICY_FULL_MEMBERS_ONLY":3,
    "POLICY_MEMBERS_ONLY":1,
    "PRIVATE_MESSAGE_POLICY_DISABLED":2,
    "PRIVATE_MESSAGE_POLICY_TYPES":[
      1,
      2
    ],
    "PRIVATE_MESSAGE_POLICY_UNLIMITED":1,
    "SELF_HOSTED":1,
    "STANDARD":3,
    "STANDARD_FREE":4,
    "STREAM_EVENTS_NOTIFICATION_TOPIC":[
      "s",
      "t",
      "r",
      "e",
      "a",
      "m",
      " ",
      "e",
      "v",
      "e",
      "n",
      "t",
      "s"
    ],
    "SUBDOMAIN_FOR_ROOT_DOMAIN":"",
    "UPGRADE_TEXT_STANDARD":[
      "A",
      "v",
      "a",
      "i",
      "l",
      "a",
      "b",
      "l",
      "e",
      " ",
      "o",
      "n",
      " ",
      "Z",
      "u",
      "l",
      "i",
      "p",
      " ",
      "S",
      "t",
      "a",
      "n",
      "d",
      "a",
      "r",
      "d",
      ".",
      " ",
      "U",
      "p",
      "g",
      "r",
      "a",
      "d",
      "e",
      " ",
      "t",
      "o",
      " ",
      "a",
      "c",
      "c",
      "e",
      "s",
      "s",
      "."
    ],
    "UPLOAD_QUOTA_LIMITED":5,
    "UPLOAD_QUOTA_STANDARD":50,
    "USER_GROUP_EDIT_POLICY_ADMINS":2,
    "USER_GROUP_EDIT_POLICY_MEMBERS":1,
    "USER_GROUP_EDIT_POLICY_TYPES":[
      1,
      2
    ],
    "VIDEO_CHAT_PROVIDERS":{
      "disabled":{
        "name":"None",
        "id":0
      },
      "jitsi_meet":{
        "name":"Jitsi Meet",
        "id":1
      },
      "zoom":{
        "name":"Zoom",
        "id":3
      },
      "big_blue_button":{
        "name":"Big Blue Button",
        "id":4
      }
    },
    "add_emoji_by_admins_only":false,
    "allow_community_topic_editing":true,
    "allow_edit_history":true,
    "allow_message_deleting":false,
    "allow_message_editing":true,
    "authentication_methods":[
      [
        "Google",
        true
      ],
      [
        "Email",
        true
      ],
      [
        "GitHub",
        true
      ],
      [
        "LDAP",
        true
      ],
      [
        "Dev",
        true
      ],
      [
        "RemoteUser",
        true
      ],
      [
        "AzureAD",
        true
      ],
      [
        "SAML",
        true
      ],
      [
        "GitLab",
        true
      ],
      [
        "Apple",
        true
      ]
    ],
    "avatar_changes_disabled":false,
    "bot_creation_policy":1,
    "create_stream_policy":1,
    "date_created":1596760684,
    "deactivated":false,
    "default_code_block_language":null,
    "default_language":"en",
    "default_twenty_four_hour_time":false,
    "description":"The Zulip development environment default organization.  It's great for testing!",
    "digest_emails_enabled":false,
    "digest_weekday":1,
    "disallow_disposable_email_addresses":true,
    "display_subdomain":"zulip",
    "email_address_visibility":3,
    "email_changes_disabled":false,
    "emails_restricted_to_domains":false,
    "first_visible_message_id":0,
    "host":"zulip.testserver",
    "icon_source":"G",
    "icon_version":1,
    "id":2,
    "inline_image_preview":true,
    "inline_url_embed_preview":false,
    "invite_by_admins_only":false,
    "invite_required":false,
    "invite_to_stream_policy":1,
    "is_zephyr_mirror_realm":false,
    "logo_source":"D",
    "logo_version":1,
    "mandatory_topics":false,
    "max_invites":100,
    "message_content_allowed_in_email_notifications":true,
    "message_content_delete_limit_seconds":600,
    "message_content_edit_limit_seconds":600,
    "message_retention_days":-1,
    "message_visibility_limit":null,
    "name":"Zulip Dev",
    "name_changes_disabled":false,
    "night_logo_source":"D",
    "night_logo_version":1,
    "notifications_stream":null,
    "notifications_stream_id":null,
    "org_type":1,
    "pk":2,
    "plan_type":1,
    "presence_disabled":false,
    "private_message_policy":1,
    "property_types":{
      "add_emoji_by_admins_only":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "allow_edit_history":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "allow_message_deleting":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "bot_creation_policy":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "create_stream_policy":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "invite_to_stream_policy":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "default_language":{

      },
      "default_twenty_four_hour_time":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "description":{

      },
      "digest_emails_enabled":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "disallow_disposable_email_addresses":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "email_address_visibility":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "email_changes_disabled":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "invite_required":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "invite_by_admins_only":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "inline_image_preview":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "inline_url_embed_preview":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "mandatory_topics":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "message_retention_days":[
        {
          "denominator":{

          },
          "imag":{

          },
          "numerator":{

          },
          "real":{

          }
        },
        {

        }
      ],
      "name":{

      },
      "name_changes_disabled":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "avatar_changes_disabled":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "emails_restricted_to_domains":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "send_welcome_emails":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "message_content_allowed_in_email_notifications":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "video_chat_provider":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "waiting_period_threshold":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "digest_weekday":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "private_message_policy":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "user_group_edit_policy":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      },
      "default_code_block_language":[
        {

        },
        {

        }
      ],
      "message_content_delete_limit_seconds":{
        "denominator":{

        },
        "imag":{

        },
        "numerator":{

        },
        "real":{

        }
      }
    },
    "send_welcome_emails":true,
    "signup_notifications_stream":null,
    "signup_notifications_stream_id":null,
    "string_id":"zulip",
    "subdomain":"zulip",
    "upload_quota_gb":null,
    "uri":"http:\/\/zulip.testserver",
    "user_group_edit_policy":1,
    "video_chat_provider":1,
    "waiting_period_threshold":0,
    "webathena_enabled":false
  },
  "realm_id":2,
  "recipient":{
    "HUDDLE":3,
    "PERSONAL":1,
    "STREAM":2,
    "id":23,
    "pk":23,
    "type":2,
    "type_id":1
  },
  "recipient_id":23,
  "rendered_description":"<p>A Scandinavian country<\/p>",
  "stream_post_policy":1
}
```

Reply:
@andersk Oh yes!. I should have use stream id instead. I'll fix that. Thanks for pointing out.

Reply:
See #16063.